File: src/combat/PlayerDamage.ts
import type Phaser from 'phaser';
import { COMBAT } from './tuning';

export class PlayerDamage {
  health: number = COMBAT.playerHealth;
  state: 'normal' | 'hit' | 'dead' = 'normal';
  invulnerabilityRemaining = 0;
  private stateRemaining = 0;

  constructor(private readonly player: Phaser.Physics.Arcade.Image) {}

  get readyToRespawn(): boolean { return this.state === 'dead' && this.stateRemaining === 0; }

  update(delta: number): void {
    this.invulnerabilityRemaining = Math.max(0, this.invulnerabilityRemaining - delta);
    this.stateRemaining = Math.max(0, this.stateRemaining - delta);
    if (this.state === 'hit' && this.stateRemaining === 0) this.state = 'normal';
  }

  receive(amount: number, sourceX: number, facing: -1 | 1): boolean {
    if (this.state !== 'normal' || this.invulnerabilityRemaining > 0 || amount <= 0) return false;
    this.health = Math.max(0, this.health - amount);
    const body = this.player.body as Phaser.Physics.Arcade.Body;
    body.maxVelocity.y = Infinity;
    if (this.health === 0) {
      this.state = 'dead';
      this.stateRemaining = COMBAT.respawnDelay;
      body.setVelocity(0, 0);
      body.enable = false;
      this.player.setVisible(false);
    } else {
      this.state = 'hit';
      this.stateRemaining = COMBAT.hitReactionDuration;
      this.invulnerabilityRemaining = COMBAT.postHitInvulnerability;
      const away = this.player.x === sourceX ? -facing : Math.sign(this.player.x - sourceX);
      body.setVelocity(away * COMBAT.receivedKnockbackHorizontal, COMBAT.receivedKnockbackVertical);
    }
    return true;
  }

  reset(): void {
    this.health = COMBAT.playerHealth;
    this.state = 'normal';
    this.stateRemaining = 0;
    this.invulnerabilityRemaining = 0;
    (this.player.body as Phaser.Physics.Arcade.Body).enable = true;
    this.player.setVisible(true);
  }
}

File: Assets/Scripts/Combat/PlayerDamage.cs
using UnityEngine;

namespace Nightblade
{
    [DisallowMultipleComponent]
    [RequireComponent(typeof(Rigidbody2D), typeof(PlayerHealth), typeof(PlayerMovement))]
    public sealed class PlayerDamage : MonoBehaviour
    {
        public enum DamageState { Normal, Hit, Dead }

        [Header("Hit response (world units and seconds)")]
        [SerializeField, Min(0f)] private float knockbackHorizontal = 4.6875f;
        [SerializeField, Min(0f)] private float knockbackVertical = 5f;
        [SerializeField, Min(0f)] private float hitReactionDuration = 0.18f;
        [SerializeField, Min(0f)] private float invulnerabilityDuration = 0.45f;

        [Header("Death")]
        [SerializeField, Min(0f)] private float respawnDelay = 0.8f;
        [SerializeField] private GameObject visualRoot;

        private Rigidbody2D body;
        private PlayerHealth health;
        private PlayerMovement movement;
        private PlayerCombat combat;
        private PlayerCharacter character;
        private float stateRemaining;

        public DamageState State { get; private set; } = DamageState.Normal;
        public float InvulnerabilityRemaining { get; private set; }
        public float StateRemaining => stateRemaining;
        public bool IsInvulnerable => InvulnerabilityRemaining > 0f;
        public bool ReadyToRespawn => State == DamageState.Dead && stateRemaining == 0f;
        public bool IsInHitReaction => State == DamageState.Hit;

        private void Awake()
        {
            body = GetComponent<Rigidbody2D>();
            health = GetComponent<PlayerHealth>();
            movement = GetComponent<PlayerMovement>();
            combat = GetComponent<PlayerCombat>();
            character = GetComponent<PlayerCharacter>();
            if (combat == null || visualRoot == null)
            {
                Debug.LogError("PlayerDamage needs player combat and a separate visual root.", this);
                enabled = false;
            }
        }

        private void Update()
        {
            InvulnerabilityRemaining = Mathf.Max(0f, InvulnerabilityRemaining - Time.deltaTime);
            stateRemaining = Mathf.Max(0f, stateRemaining - Time.deltaTime);
            if (State != DamageState.Hit || stateRemaining > 0f) return;

            State = DamageState.Normal;
            bool runActive = character == null || character.RunInProgress;
            movement.SetControlEnabled(runActive);
            combat.SetControlEnabled(runActive);
        }

        public bool ReceiveDamage(int amount, float sourceX)
        {
            if (GetComponent<FireCharacterGameplay>()?.IsShieldActive == true ||
                GetComponent<WaterCharacterGameplay>()?.IsShieldActive == true ||
                GetComponent<AirCharacterGameplay>()?.IsShieldActive == true) return false;
            if (State != DamageState.Normal || IsInvulnerable || amount <= 0) return false;
            if (!health.TakeDamage(amount)) return false;

            movement.SetControlEnabled(false);
            combat.SetControlEnabled(false);
            if (health.IsDepleted)
            {
                EnterDeadState();
                return true;
            }

            State = DamageState.Hit;
            stateRemaining = hitReactionDuration;
            InvulnerabilityRemaining = invulnerabilityDuration;
            int away = Mathf.Approximately(body.position.x, sourceX)
                ? -combat.Facing
                : body.position.x > sourceX ? 1 : -1;
            body.linearVelocity = new Vector2(away * knockbackHorizontal, knockbackVertical);
            return true;
        }

        public void RespawnAt(Vector2 worldPosition)
        {
            health.RestoreFull();
            State = DamageState.Normal;
            stateRemaining = 0f;
            InvulnerabilityRemaining = 0f;
            visualRoot.SetActive(true);
            movement.ResetMovement(worldPosition);
            combat.ResetCombat();
        }

        private void EnterDeadState()
        {
            State = DamageState.Dead;
            stateRemaining = respawnDelay;
            InvulnerabilityRemaining = 0f;
            body.linearVelocity = Vector2.zero;
            body.simulated = false;
            visualRoot.SetActive(false);
        }
    }
}
